File: app/src/main/java/com/example/music/ui/home/HomeViewModel.java
package com.example.music.ui.home;

import android.app.Application;
import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import com.example.music.data.api.RetrofitClient;
import com.example.music.data.db.MusicDatabase;
import com.example.music.data.db.TrackDao;
import com.example.music.data.model.Album;
import com.example.music.data.model.Artist;
import com.example.music.data.model.DataResponse;
import com.example.music.data.model.Track;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import com.example.music.data.repository.MusicRepository;
import com.example.music.data.api.SupabaseApi;

public class HomeViewModel extends AndroidViewModel {

    private final TrackDao trackDao;
    private final ExecutorService executorService;
    private final MutableLiveData<String> error = new MutableLiveData<>();
    private final MusicRepository musicRepository;

    public HomeViewModel(@NonNull Application application) {
        super(application);
        MusicDatabase db = MusicDatabase.getInstance(application);
        trackDao = db.trackDao();
        executorService = Executors.newSingleThreadExecutor();
        musicRepository = new MusicRepository(RetrofitClient.getApi());
    }

    public LiveData<List<Track>> getTracks() {
        return trackDao.getAllTracks();
    }

    public LiveData<List<Album>> getAlbums() {
        // We can derive albums from tracks in the fragment or create a separate DAO
        // query
        // For simplicity, let's derive them in the fragment or use a MediatorLiveData
        MutableLiveData<List<Album>> albums = new MutableLiveData<>();
        getTracks().observeForever(trackList -> {
            if (trackList != null) {
                java.util.Set<String> albumIds = new java.util.HashSet<>();
                List<Album> albumList = new java.util.ArrayList<>();
                for (Track track : trackList) {
                    Album album = track.getAlbum();
                    if (album != null && !albumIds.contains(String.valueOf(album.getId()))) {
                        albumIds.add(String.valueOf(album.getId()));
                        albumList.add(album);
                    }
                }
                albums.postValue(albumList);
            }
        });
        return albums;
    }

    public LiveData<String> getError() {
        return error;
    }

    public void fetchCharts() {
        musicRepository.getChartTracks().observeForever(tracks -> {
            if (tracks != null) {
                executorService.execute(() -> {
                    trackDao.insertTracks(tracks);
                });
            } else {
                error.postValue("Failed to load cloud tracks from Supabase");
            }
        });
    }
}

File: app/src/main/java/com/example/music/ui/home/HomeViewModel.kt
package com.example.music.ui.home

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.example.music.data.model.DeezerAlbum
import com.example.music.data.model.Track
import com.example.music.data.repository.AudiusRepository
import com.example.music.data.repository.DeezerRepository
import com.example.music.data.repository.ItunesRepository
import com.example.music.data.repository.JamendoRepository
import com.example.music.data.repository.SupabaseRepository
import kotlinx.coroutines.async
import kotlinx.coroutines.launch

class HomeViewModel(application: Application) : AndroidViewModel(application) {

    private val deezerRepository = DeezerRepository()
    private val itunesRepository = ItunesRepository()
    private val jamendoRepository = JamendoRepository()
    private val audiusRepository = AudiusRepository()
    private val supabaseRepository = SupabaseRepository()

    // Deezer — chart tracks (grid + "Trending Now" row)
    private val _tracks = MutableLiveData<List<Track>>()
    val tracks: LiveData<List<Track>> get() = _tracks

    // Deezer — chart albums
    private val _deezerAlbums = MutableLiveData<List<DeezerAlbum>>()
    val deezerAlbums: LiveData<List<DeezerAlbum>> get() = _deezerAlbums

    // iTunes — top hits row
    private val _itunesTracks = MutableLiveData<List<Track>>()
    val itunesTracks: LiveData<List<Track>> get() = _itunesTracks

    // Jamendo — full free songs row
    private val _jamendoTracks = MutableLiveData<List<Track>>()
    val jamendoTracks: LiveData<List<Track>> get() = _jamendoTracks

    private val _hindiEnglishTracks = MutableLiveData<List<Track>>()
    val hindiEnglishTracks: LiveData<List<Track>> get() = _hindiEnglishTracks

    private val _audiusTracks = MutableLiveData<List<Track>>()
    val audiusTracks: LiveData<List<Track>> get() = _audiusTracks

    // Supabase — saved/uploaded tracks
    private val _savedTracks = MutableLiveData<List<Track>>()
    val savedTracks: LiveData<List<Track>> get() = _savedTracks

    private val _isLoading = MutableLiveData(false)
    val isLoading: LiveData<Boolean> get() = _isLoading

    private val _error = MutableLiveData<String?>()
    val error: LiveData<String?> get() = _error

    fun fetchCharts() {
        viewModelScope.launch {
            _isLoading.value = true
            _error.value = null

            // Fire all 4 network calls in parallel
            val deezerTracksDeferred  = async { deezerRepository.getChartTracks(50) }
            val deezerAlbumsDeferred  = async { deezerRepository.getChartAlbums(20) }
            val itunesDeferred        = async { itunesRepository.getTopHits(30) }
            val jamendoDeferred       = async { jamendoRepository.getPopularTracks(20) }
            val audiusDeferred        = async { audiusRepository.getTrendingTracks(30) }
            val hindiDeferred         = async { itunesRepository.searchTracks("hindi bollywood songs", 20) }
            val englishDeferred       = async { itunesRepository.searchTracks("english pop hits", 20) }

            // Deezer tracks
            deezerTracksDeferred.await()
                .onSuccess { _tracks.postValue(it) }
                .onFailure { _error.postValue("Deezer: ${it.localizedMessage}") }

            // Deezer albums
            deezerAlbumsDeferred.await()
                .onSuccess { _deezerAlbums.postValue(it) }

            // iTunes top hits
            itunesDeferred.await()
                .onSuccess { _itunesTracks.postValue(it) }
                .onFailure { _error.postValue("iTunes: ${it.localizedMessage}") }

            // Jamendo free full songs
            jamendoDeferred.await()
                .onSuccess { _jamendoTracks.postValue(it) }
                .onFailure { _error.postValue("Jamendo: ${it.localizedMessage}") }

            audiusDeferred.await()
                .onSuccess { _audiusTracks.postValue(it) }
                .onFailure { _error.postValue("Audius: ${it.localizedMessage}") }

            val hindiTracks = hindiDeferred.await().getOrElse { emptyList() }
            val englishTracks = englishDeferred.await().getOrElse { emptyList() }
            _hindiEnglishTracks.postValue(interleaveAndDeduplicate(hindiTracks, englishTracks))

            _isLoading.value = false
        }
    }

    private fun interleaveAndDeduplicate(first: List<Track>, second: List<Track>): List<Track> {
        val mixed = buildList {
            val a = first.iterator()
            val b = second.iterator()
            while (a.hasNext() || b.hasNext()) {
                if (a.hasNext()) add(a.next())
                if (b.hasNext()) add(b.next())
            }
        }
        val seen = mutableSetOf<String>()
        return mixed.filter { track ->
            val key = "${track.title.lowercase().trim()}|${track.artist.name.lowercase().trim()}"
            seen.add(key)
        }
    }

    fun fetchSavedTracks() {
        viewModelScope.launch {
            supabaseRepository.getSavedTracks()
                .onSuccess { _savedTracks.postValue(it) }
        }
    }
}
